core/src/torii/routing.rs

```diff
@@ -161,16 +161,84 @@ async fn stream_blocks<W: WorldTrait>(
-    let mut consumer = Consumer::new(stream).await?;
+mod subscription {
+    //! This module contains `handle_subscription()` function and other stuff for its implementation
+
+    use super::*;
+    use crate::{event, stream};
+
+    /// Type for any error during subscription handling
+    #[derive(thiserror::Error, Debug)]
+    enum Error {
+        /// Event consuming error
+        #[error("Event consuming error: {0}")]
+        Consumer(Box<event::Error>),
+        /// Event receiving error
+        #[error("Event receiving error: {0}")]
+        Event(#[from] tokio::sync::broadcast::error::RecvError),
+        /// Error from provided websocket
+        #[error("WebSocket error: {0}")]
+        WebSocket(#[from] warp::Error),
+        /// Error, indicating that `Close` message was received
+        #[error("`Close` message received")]
+        CloseMessage,
+    }
 
-    loop {
-        let event = events.recv().await?;
+    type StreamError = stream::Error<<WebSocket as Stream<VersionedEventSubscriberMessage>>::Err>;
+
+    impl From<event::Error> for Error {
+        fn from(error: event::Error) -> Self {
+            match error {
+                event::Error::Stream(StreamError::CloseMessage) => Self::CloseMessage,
+                error => Self::Consumer(Box::new(error)),
+            }
+        }
+    }
+
+    type Result<T> = core::result::Result<T, Error>;
+
+    /// Handle subscription request
+    ///
+    /// Subscribes `stream` for `events` filtered by filter that is received through the `stream`
+    ///
+    /// There should be a [`warp::filters::ws::Message::close()`] message to end subscription
+    #[iroha_futures::telemetry_future]
+    pub async fn handle_subscription(events: EventsSender, stream: WebSocket) -> eyre::Result<()> {
+        match subscribe_forever(events, stream).await {
+            Ok(()) | Err(Error::CloseMessage) => Ok(()),
+            Err(err) => Err(err.into()),
+        }
+    }
 
-        iroha_logger::trace!(?event);
-        consumer.consume(event).await?;
+    /// Make endless `stream` subscription for `events`
+    ///
+    /// Ideally should return `Result<!>` cause it either runs forever either returns `Err` variant
+    async fn subscribe_forever(events: EventsSender, stream: WebSocket) -> Result<()> {
+        let mut events = events.subscribe();
+        let mut consumer = Consumer::new(stream).await?;
+
+        loop {
+            tokio::select! {
+                // This branch catches `Close` ans unexpected messages
+                message_opt = consumer.stream_mut().next() => {
+                    if let Some(message) = message_opt {
+                        let message = message?;
+                        if message.is_close() {
+                            return Err(Error::CloseMessage);
+                        }
+                        iroha_logger::trace!("Unexpected message received: {:?}", message);
+                    }
+                }
```

Comment:
this is getting better, but you don't need this `select!` mechanics at all if you're relying on the returned error value. Just let any of the calls to consumer bubble up the error to where you can handle it

Reply:
I think, that won't work for case when no events are received (or passed filter) but client want to close connection. In that case client has to wait until first `recv()` on server side that will happen only when server get new event, send it and call `recv` excpecting`EventReceived` message

Reply:
if client wants to close connection, client can close connection regardless of what server is doing

Reply:
Yeah, but that wouldn't be clean closing. From #1195 issue:

> For now, the js client will force connections to terminate.

I think the reason of this problem is that [Closing handshake](https://datatracker.ietf.org/doc/html/draft-ietf-hybi-thewebsocketprotocol#page-10) requires response with the same `Close` message.

Interesting, that `warp` WebSocket seems to do this response automatically, so I don't have to do anything. However I still has to call `recv()` (or `next()`) to retrieve that message. I've found this out during tests, but additional tests with some real js client would be good

Reply:
I thought that closing connection abbruptly on the client side would produce an error on the server, not on the client

Yes, warp websocket handles close message automatically. I think that is documented

Reply:
After talk in DM we decided to resolve this conversation and leave everything like it is

Reply:
What is that conclusion?

Reply:
`select!` is needed because while waiting on events a close message can be received on the websocket and not get handled until new event is received

Reply:
BTW what is the reason we don't use [`WebSocket::close`](https://docs.rs/warp/latest/warp/filters/ws/struct.WebSocket.html#method.close) ?

Reply:
Cause we don't have to explicitly close socket like this. It's somehow being closed automatically by `warp` (if you try to send something after receiving `Close` message you'll get error with message `Connection closed normally`).
My test is working without `WebSocket::close()` on server side. May be there is any case which can catch this, but I don't know it

Reply:
This implies a contradiction with the warp docs. Worth reporting
> Close messages need to be handled explicitly

Reply:
I'm gonna create minimal example for that soon and check it. If I reproduce this behaviour, then I'll report them